aten/src/ATen/native/cuda/PersistentSoftmax.cuh

```diff
@@ -0,0 +1,385 @@
+        }
+    }
+
+    // reduction max_value
+    #pragma unroll
+    for (int offset = WARP_SIZE / 2; offset > 0; offset /= 2) {
+        acc_t val[WARP_BATCH];
+        #pragma unroll
+        for (int i = 0;i < WARP_BATCH;++i) {
+            val[i] = WARP_SHFL_XOR(max_value[i], offset, WARP_SIZE, FULL_MASK);
+        }
+        #pragma unroll
+        for (int i = 0;i < WARP_BATCH;++i) {
+            max_value[i] = max_value[i] > val[i] ? max_value[i] : val[i];
+        }
+    }
+
+    // compute local sum
+    acc_t sum[WARP_BATCH] { 0.0f };
+
+    #pragma unroll
+    for (int i = 0;i < WARP_BATCH;++i) {
+        for (int it = 0;it < WARP_ITERATIONS;++it) {
+            if (is_log_softmax) {
+              sum[i] += std::exp(elements[i][it] - max_value[i]);
+            } else {
+              elements[i][it] = std::exp(elements[i][it] - max_value[i]);
+              sum[i] += elements[i][it];
+            }
+        }
+    }
+
+    // reduction sum
+    #pragma unroll
+    for (int offset = WARP_SIZE / 2; offset > 0; offset /= 2) {
+        #pragma unroll
+        for (int i = 0;i < WARP_BATCH;++i) {
+            sum[i] += WARP_SHFL_XOR(sum[i], offset, WARP_SIZE, FULL_MASK);
+        }
+    }
+
+    // store result
+    #pragma unroll
+    for (int i = 0;i < WARP_BATCH;++i) {
+        if (i >= local_batches)
+            break;
+        if (is_log_softmax) sum[i] = max_value[i] + std::log(sum[i]);
+        #pragma unroll
+        for (int it = 0;it < WARP_ITERATIONS;it += 1) {
+            int element_index = local_idx + it * WARP_SIZE;
+            if (element_index < element_count) {
+                if (is_log_softmax) {
+                    dst[i*element_count+it*WARP_SIZE] = elements[i][it] - sum[i];
+                } else {
+                    dst[i*element_count+it*WARP_SIZE] = elements[i][it] / sum[i];
+                }
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+
+// WARP_BATCH number of batches.
+// WARP_ITERATOINS The number of iterations required for one warp to iterate over all data.
+// WARP_SIZE number of elements working on a single batch, has to be a power of two.
+template <typename input_t, typename output_t>
+using softmax_forward_func = void(*)(output_t *dst, const input_t *src, int batch_size, int stride, int element_count);
+
+template <typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool warp_softmax_kernel(int log2_elements, int &warp_size, int &batches_per_warp, softmax_forward_func<input_t, output_t> &kernel) {
+    // determine size of a warp
+    const int next_power_of_two = 1 << log2_elements;
+    warp_size = (next_power_of_two < 32) ? next_power_of_two : 32;
+
+    // determine how many batches a warp should process.
+    batches_per_warp = (next_power_of_two <= 128) ? 2 : 1;
+
+    switch (log2_elements) {
+    case 0: // 1
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,1, is_log_softmax>;
```

Comment:
It looks to me like all the template arguments can be constexprs in the kernel, and the only real template argument is `next_power_of_2`. E.g.
```
WARP_SIZE = (np2 <  = 32) ? np2 : 32;
WARP_ITERATIONS = np2/WARP_SIZE;
WARP_BATCH = yet another ternary operator
```
This will make dispatch for forward and backward cleaner.